### PR TITLE
Adding a getter and setter for cache registry

### PR DESCRIPTION
**Pull request by burzum**

### CODE REVIEW

**Review by Theaxiom — APPROVED** [review]

File: src/Cache/Cache.php [hunk 1]
```diff
@@ -108,22 +108,45 @@ class Cache
...
+        }
+
+        return static::$_registry;
+    }
+
+    /**
+     * Returns the Cache Registry instance used for creating and using cache adapters.
+     * Also allows for injecting of a new registry instance.
+     *
+     * @param \Cake\Core\ObjectRegistry|null $registry Injectable registry object.
```
Comment (cleptric) [inline src/Cache/Cache.php]: `null` is obsolete now

File: src/Cache/Cache.php [hunk 2]
```diff
@@ -108,22 +108,45 @@ class Cache
...
+    public static function setRegistry(ObjectRegistry $registry)
+    {
+        static::$_registry = $registry;
+    }
+
+    /**
+     * Returns the Cache Registry instance used for creating and using cache adapters.
      * Also allows for injecting of a new registry instance.
      *
+     * @deprecated Use getRegistry() and setRegistry() instead.
```
Comment (cleptric) [inline src/Cache/Cache.php]: Could we add the version number?

Comment (burzum) [inline src/Cache/Cache.php]: OK, but why do you need it? I don't think *any* other deprecated method has the version number included.

Comment (cleptric) [inline src/Cache/Cache.php]: That's not true. A quick search for `@deprecated` inside the core reveals that basically all methods have a version number attached to the annotation.
Main reason here is to be consistent :)

Comment (Theaxiom) [inline src/Cache/Cache.php]: If that's the case, we should also raise an issue against any deprecation notices which omit the deprecated version number.

Comment (burzum) [inline src/Cache/Cache.php]: Yes, I'm not 100% sure but I haven't seen any version in these lines before. Were they added recently?

Comment (cleptric) [inline src/Cache/Cache.php]: We normally just add them when we deprecate stuff.

@Theaxiom There are just a few methods that missing the version. I guess it's a little bit time consuming to scrape through the history to find the right version. A PR addressing this is of course welcomed at any time 😀

File: src/Cache/Cache.php [hunk 3]
```diff
@@ -108,22 +108,45 @@ class Cache
...
+    {
+        if (!static::$_registry) {
+            static::$_registry = new CacheRegistry();
+        }
+
+        return static::$_registry;
+    }
+
+    /**
+     * Returns the Cache Registry instance used for creating and using cache adapters.
```
Comment (markstory) [inline src/Cache/Cache.php]: But it doesn't return anything..

**Review by markstory — CHANGES_REQUESTED** [review]
Would be good to have some tests for the new methods.

File: src/Cache/Cache.php (hunk 3 above)
Comment (ravage84) [inline src/Cache/Cache.php]: Probably Copy Pasta error.

File: src/Cache/Cache.php [hunk 4]
```diff
@@ -108,22 +108,45 @@ class Cache
...
+    public static function setRegistry(ObjectRegistry $registry)
+    {
+        static::$_registry = $registry;
+    }
+
+    /**
+     * Returns the Cache Registry instance used for creating and using cache adapters.
+     * Also allows for injecting of a new registry instance.
+     *
+     * @deprecated Deprecated since 3.5. Use getRegistry() and setRegistry() instead.
```
Comment (ravage84) [inline src/Cache/Cache.php]: Move below return.

**Review by ravage84 — CHANGES_REQUESTED** [review]

**Comment by burzum** [comment]
@markstory I've added tests, please review again.

File: src/Cache/Cache.php [hunk 5]
```diff
@@ -165,16 +189,31 @@ protected static function _buildEngine($name)
...
+    /**
+     * Fetch the engine attached to a specific configuration name.
+     *
+     * If the cache engine & configuration are missing an error will be
+     * triggered.
+     *
+     * @param string $config The configuration name you want an engine for.
+     * @return \Cake\Cache\CacheEngine When caching is disabled a null engine will be returned.
+     */
+    public static function getEngine($config)
```
Comment (saeideng) [inline src/Cache/Cache.php]: we have not setter method for `Engine() ` , so we no need to `getEngine()`

Comment (markstory) [inline src/Cache/Cache.php]: Agreed.

**Pull request merged**